### PR TITLE
perf(spec-parser): update to support type b ai-plugin

### PR DESCRIPTION
**Pull request by SLdragon**
Example:

Code sample:
```ts
  const testFolder = path.resolve("../", path.dirname(__dirname), "test-material");
  const manifest = testFolder + "/manifest.json";
  const spec = testFolder + "/petstore.yaml";
  const parser = new SpecParser(spec, {
    allowMultipleParameters: true,
    isCopilot: true,
  });
  const list = await parser.list();
  const validate = await parser.validate();
  const apis = list.map((item) => item.api);
  const result = await parser.generateForCopilot(manifest, apis, testFolder + "/" + "result.yaml", "ai-plugin.json");
```

OpenAPI spec file:
```yaml
openapi: 3.0.0
info:
  title: Phone Service
  description: >-
    http service to operate phone via http request, such as send message, get
    contacts, etc
  version: 1.0.2
servers:
  - url: https://dcg.microsoft.com
paths:
  /DeviceDataConnector/devices/default/contacts:
    get:
      operationId: getContactList
      summary: Get contact list from device.
      description: Get contact list from device.
      parameters:
        - name: top
          in: query
          description: The top number of contacts to return, default is 100
          required: false
          schema:
            type: string
        - name: filter
          in: query
          description: >-
            The filter used to filter the contacts, currently support filter by
            name, company and location, such as `filter=name has $contactName`,
            also it supports `and` and `or` operator, such as `filter=name has
            $contactName or company has $company`
          required: false
          schema:
            type: string
        - name: name
          in: query
          description: The parameter to filter the contacts by name, such as `name=Tom`
          required: false
          schema:
            type: string
        - name: company
          in: query
          description: >-
            The parameter to filter the contacts by company, such as
            `company=Microsoft`
          required: false
          schema:
            type: string
      responses:
        '200':
         ...
```

Generated manifest file:
```json
{
  "$schema": "https://raw.githubusercontent.com/OfficeDev/microsoft-teams-app-schema/preview/DevPreview/MicrosoftTeams.schema.json",
  "manifestVersion": "devPreview",
  "version": "1.0.0",
  "id": "${{TEAMS_APP_ID}}",
  "packageName": "com.microsoft.teams.extension",
  "developer": {
    "name": "Teams App, Inc.",
    "websiteUrl": "https://www.example.com",
    "privacyUrl": "https://www.example.com/termofuse",
    "termsOfUseUrl": "https://www.example.com/privacy"
  },
  "icons": {
    "color": "color.png",
    "outline": "outline.png"
  },
  "name": {
    "short": "Phone Service",
    "full": "Phone Service"
  },
  "description": {
    "short": "Phone Service",
    "full": "http service to operate phone via http request, such as send message, get contacts, etc"
  },
  "accentColor": "#FFFFFF",
  "composeExtensions": [],
  "permissions": [
    "identity",
    "messageTeamMembers"
  ],
  "validDomains": [],
  "apiPlugins": [
    {
      "pluginFile": "ai-plugin.json"
    }
  ]
}

```

Generated ai-plugin file:
```json
{
  "schema_version": "v2",
  "name_for_human": "Phone Service",
  "description_for_human": "http service to operate phone via http request, such as send message, get contacts, etc",
  "functions": [
    {
      "name": "getContactList",
      "description": "Get contact list from device.",
      "parameters": {
        "type": "object",
        "properties": {
          "top": {
            "type": "string",
            "description": "The top number of contacts to return, default is 100"
          },
          "filter": {
            "type": "string",
            "description": "The filter used to filter the contacts, currently support filter by name, company and location, such as `filter=name has $contactName`, also it supports `and` and `or` operator, such as `filter=name has $contactName or company has $company`"
          },
          "name": {
            "type": "string",
            "description": "The parameter to filter the contacts by name, such as `name=Tom`"
          },
          "company": {
            "type": "string",
            "description": "The parameter to filter the contacts by company, such as `company=Microsoft`"
          }
        },
        "required": []
      },
      "states": {
        "reasoning": {
          "description": "",
          "instructions": []
        },
        "responding": {
          "description": "",
          "instructions": []
        }
      }
    }
  ],
  "runtimes": [
    {
      "type": "OpenApi",
      "auth": {
        "type": "none"
      },
      "spec": {
        "url": "result.yaml"
      },
      "run_for_functions": [
        "getContactList"
      ]
    }
  ]
}
```

### CODE REVIEW

File: packages/spec-parser/src/specParser.browser.ts [hunk 1]
```diff
@@ -163,6 +165,21 @@ export class SpecParser {
...
+   * @param filter An array of strings that represent the filters to apply when generating the artifacts. If filter is empty, it would process nothing.
+   * @param outputSpecPath File path of the new OpenAPI specification file to generate. If not specified or empty, no spec file will be generated.
+   */
+  // eslint-disable-next-line @typescript-eslint/require-await
+  async generateForCopilot(
+    manifestPath: string,
+    filter: string[],
+    outputSpecPath: string,
+    signal?: AbortSignal
+  ): Promise<GenerateResult> {
```
Comment (yuqizhou77) [inline packages/spec-parser/src/specParser.browser.ts:180]: should we also make ai-plugin.json as an input parameter? I think user may customize the file location when they want to use the codelens to add a new API?

File: packages/spec-parser/src/manifestUpdater.ts [hunk 2]
```diff
@@ -13,9 +13,193 @@ import {
...
+                }
+              }
+
+              const funcObj: FunctionObject = {
+                name: operationId,
+                description: description,
+                parameters: parameters,
+                states: {
+                  reasoning: {
+                    description: ConstantString.ReasoningDescription,
```
Comment (yuqizhou77) [inline packages/spec-parser/src/manifestUpdater.ts:163]: kind of curious why we use constant string here.  sample value: https://spec-hub.azurewebsites.net/specifications/PluginManifest-draft.html#example-7 Should we leave it empty if we are not able to generate it?

Comment (SLdragon) [reply packages/spec-parser/src/manifestUpdater.ts:163]: Sure, we leave them empty

File: packages/spec-parser/src/specParser.browser.ts (hunk 1 above)
Comment (SLdragon) [reply packages/spec-parser/src/specParser.browser.ts:180]: Sure

File: packages/spec-parser/src/specParser.browser.ts [hunk 3]
```diff
@@ -163,6 +165,22 @@ export class SpecParser {
...
+   * @param manifestPath A file path of the Teams app manifest file to update.
+   * @param filter An array of strings that represent the filters to apply when generating the artifacts. If filter is empty, it would process nothing.
+   * @param outputSpecPath File path of the new OpenAPI specification file to generate. If not specified or empty, no spec file will be generated.
+   */
+  // eslint-disable-next-line @typescript-eslint/require-await
+  async generateForCopilot(
+    manifestPath: string,
+    filter: string[],
+    outputSpecPath: string,
+    apiPluginFileName: string,
```
Comment (yuqizhou77) [inline packages/spec-parser/src/specParser.browser.ts:179]: just want to confirm if you hope to pass the file name. might  be better if it is path?

Comment (SLdragon) [reply packages/spec-parser/src/specParser.browser.ts:179]: Will update to use path instead of name

**Review by yuqizhou77 — APPROVED** [review]

**Review by KennethBWSong — APPROVED** [review]

**Review by adashen — APPROVED** [review]

**Pull request merged**